### ISSUE TITLE
[BUG] Error parsing JSON:'NoneType' object has no attribute 'chat' on MAC

### ISSUE THREAD

**Issue by wanghaojin** [OthersideAI/self-operating-computer#128]
To test the program,input:"Go to Youtube.com and play a video."
Here's the error reproted:
"[Self-Operating Computer][Error] something went wrong :(
[Self-Operating Computer][Error] AI response
Failed take action after looking at the screenshot"
<img width="466" alt="2024-01-09 11 27 09" src="https://github.com/OthersideAI/self-operating-computer/assets/56794303/6f078fc6-0aef-4689-ab72-50bac1f9d0e6">

**Comment by michaelhhogue** [comment]
@wanghaojin Were you using `gpt-4-vision-preview` or `gemini-pro`?

**Comment by SlinkyFive30012** [comment]
@michaelhhogue I am having that same problem but on Windows 11.

**Comment by michaelhhogue** [comment]
@SlinkyFive30012 GPT-4v or Gemini Pro?

**Comment by SlinkyFive30012** [comment]
@michaelhhogue GPT-4

**Comment by michaelhhogue** [comment]
@SlinkyFive30012 Got it. I'll look into this in a little bit and get back to you.

**Comment by SlinkyFive30012** [comment]
@michaelhhogue Ok

**Comment by joshbickett** [comment]
Ok, there may be a prompt issue. I'll look closer as well

**Comment by SlinkyFive30012** [comment]
@michaelhhogue Ok

**Comment by joshbickett** [comment]
I'm having trouble recreating. It doesn't appear to be an actual JSON issue. I fixed that `print` statement because it is possible to have errors that are not JSON related that show up there. The issue is that a variable was `None` that was not expected, but which one yet

**Comment by SlinkyFive30012** [comment]
@joshbickett So what must I do to fix this issue?

**Comment by joshbickett** [comment]
If you try it again do you get the same issue or was it possibly a one time issue?

**Comment by michaelhhogue** [comment]
@joshbickett Could it be that the client is `None` because the OpenAI client initialization function is returning `None`?

```
def initialize_openai_client(self):
        """
        Initializes and returns an OpenAI client with the configured API key.

        Returns:
            OpenAI or None: An instance of the OpenAI client if the API key is provided, else None.
        """
        if self.openai_api_key:
            client = OpenAI()
            client.api_key = self.openai_api_key
            client.base_url = os.getenv("OPENAI_API_BASE_URL", client.base_url)
            return client
        return None
```

In this case, it would be due to the OpenAI key not being defined.

**Comment by joshbickett** [comment]
@michaelhhogue, good point. It looks like it is 'none' if there is no OpenAI key provided. 

So that's is what I am wondering about now. Is there a key in the Env for openAI?

**Comment by tzayuan** [comment]
same error, using GPT-4 keys

**Comment by joshbickett** [comment]
Can you try `pip show openai`? It should be on `version >= 1.2`
```
josh@Joshs-MBP self-operating-computer % pip show openai
Name: openai
Version: 1.2.3
```

**Comment by nemzyx** [comment]
@joshbickett 
```
(SOC) λ  self-operating-computer pip show openai
Name: openai
Version: 1.2.3
```
👍✅

I was initially trying to debug this repo, but i couldn't even get print statements to work so i gave up quickly.

**Comment by tzayuan** [comment]
@joshbickett 

<img width="555" alt="image" src="https://github.com/OthersideAI/self-operating-computer/assets/25738822/6da42115-47e0-4490-9ca0-a43458e00fd7">

**Comment by joshbickett** [comment]
After looking closer I think I found the issue and I am fixing it for the next version. It appears the OpenAI key validation step is failing. If you're having this issue can you double check your `.env` to see if it says `your-key-here`? I think if you're having this issue your OpenAI key may still be set to `your-key-here` and not your actual OpenAI key

```OPENAI_API_KEY='your-key-here'```

**Comment by nemzyx** [comment]
@joshbickett no i definitely set the `OPENAI_API_KEY ` to a valid obtained one. I can use the same key in other projects.

I just checked again, even trying to add quotes around the key just to be safe. Still get the same error...
Do you have any idea why i'm not able to print anyway? It's hard to debug without being able to do that.

**Comment by Cl4ySmith** [comment]
I'm not sure what the cause of this problem is, but there are two things I've tested. If you get an error like `Error parsing JSON:'NoneType' object has no attribute 'chat'`, try adding `OPENAI_API_BASE_URL` to the environment variable, set its value to `https://api.openai.com`. Close the terminal and reopen it to run `operate`. If this doesn't work, there is another way, which I'll mention below. 

Another thing, the reason why modifying the code in the `self-operating-computer\operate` directory doesn't work is because `operate` actually runs `operate.exe` in the `Scripts` directory of the python environment, which can be found by executing `where operate` in the environment terminal. The `operate.exe` will call the code in the `<YourPythonEnv>\Lib\site-packages\operate` directory, and you can get the debugging information by adding a print statement to the code in this directory. In the figure below, `AAAAA` is output by the test print statement. 

The other solution to the above error is to run `operate` in the ``<YourPythonEnv>\Lib\site-packages`` directory, which may report other errors, such as prompting for a missing file, and copy the files in the `self-operating-computer\operate` directory into the `<YourPythonEnv>\Lib\site-packages\operate` can solve the missing file errors.
![image](https://github.com/OthersideAI/self-operating-computer/assets/80500763/a2f0ccb4-8ab7-48dc-8365-3aa84270ff68)

**Comment by Cl4ySmith** [comment]
Theoretically adding the `OPENAI_API_KEY` environment variable would work, I made a mistake, since I'm using a multi-tabbed terminal, and I've just realized that closing only the tabs still doesn't reset the environment variable, i.e. I still have `OPENAI_API_KEY` in my environment variable when I'm following the second scenario. If I reopen the terminal completely without this environment variable, I still get the error `Error parsing JSON:'NoneType' object has no attribute 'chat'`. I also found that changing the `OPENAI_API_BASE_URL` environment variable doesn't work, nor does changing the url in actions.py. The real url used is line 103 of _client.py in the lib directory of openai `base_url = f "https://api. openai.com/v1"`.

**Comment by joshbickett** [comment]
> Do you have any idea why i'm not able to print anyway?

@nemzyx After changing the code run `pip install .` which will compile it and then you can run `operate` with your `print` statements. 

Sorry we've not been able to resolve this issue for you. I made changes to the initializing OpenAI library that may fix the problem, but not sure. If you want to `git pull` and try the update. I realized that if it fails to initialize then it returns `None` and sets the `client` to `None`, but I don't know why it is failing for you. I now initialize it later right before the API call so that may fix it

**Comment by joshbickett** [comment]
@Hugo-PureJ thank you for sharing. It seems this issue may be more wide spread then I realized. I'll take a closer look at resolving it

**Comment by abdullahjamill** [comment]
> same problem already install openai and set api key in .env but still occur 'NonType' problem ![image](https://private-user-images.githubusercontent.com/67493384/297310439-7eacdf64-8cd9-4d85-9ae7-f9f5aaa05c82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2MDMwMzIsIm5iZiI6MTcwNTYwMjczMiwicGF0aCI6Ii82NzQ5MzM4NC8yOTczMTA0MzktN2VhY2RmNjQtOGNkOS00ZDg1LTlhZTctZjlmNWFhYTA1YzgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDE4MzIxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwYmUyOGY4NWI0NGZhYjQ5ZTBkOGQ1ZDE3NTE2MmRkMzc2ODIyNTY2M2I4YWRmZjkzNjYxMzgwZGFmNmEzNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.khfazG90ASzS9UIu3wBkVoRYVPT6bVz8WsEE02dnSW0)

Same here

**Comment by ubaiidullaah** [comment]
Sa

> same problem already install openai and set api key in .env but still occur 'NonType' problem ![image](https://private-user-images.githubusercontent.com/67493384/297310439-7eacdf64-8cd9-4d85-9ae7-f9f5aaa05c82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2MDM4NzEsIm5iZiI6MTcwNTYwMzU3MSwicGF0aCI6Ii82NzQ5MzM4NC8yOTczMTA0MzktN2VhY2RmNjQtOGNkOS00ZDg1LTlhZTctZjlmNWFhYTA1YzgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDE4NDYxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiNGI4NTQ1MzljOGRmMjMxNzIzMGFlOTE1N2IxMDFmYTVlNTI1NGZkZDdiYTc2MmIxMDFkMDM2ZDE0ZWRhYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.F9P10_nhareeylSKCUzojmGHAh5-C5IWV1v6FcW7RaA)

Same here, but evaluate.py is working fine.

**Comment by joshbickett** [comment]
@Hugo-PureJ @wanghaojin @SlinkyFive30012 @nemzyx @ubaiidullaah @Cl4ySmith Ok, so I think I fixed the issue in `version 1.2.5` that I just published. I think it may have related to a few bugs, but the largest being that we were assigning the `api_key` to the OpenAI `client` after it was instantiated. I fixed it with the code below and also added a new screen in after `operate`. If the key does not exist after `operate` it will prompt the user for one. 
```
client = OpenAI(
            api_key=api_key,
        )
```
![image](https://github.com/OthersideAI/self-operating-computer/assets/42594239/012520c6-d869-4b29-95bf-7bd02ed5f40e)

Please let me know if this fixes your issue. Thanks you!

**Comment by joshbickett** [comment]
@Hugo-PureJ @wanghaojin, etc. I'm going to close this ticket for now since I've not heard from you on whether it is fixed and I am not able to recreate. If someone still has this issue, please reopen this ticket or another one
